sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoderSuite.scala

```diff
@@ -308,6 +309,26 @@ class ExpressionEncoderSuite extends CodegenInterpretedPlanTest with AnalysisTes
   encodeDecodeTest(Option("abc"), "option of string")
   encodeDecodeTest(Option.empty[String], "empty option of string")
 
+  encodeDecodeTest(
```

Comment:
I tried and the test case failed:
```
Can't compare maps!
org.apache.avro.AvroRuntimeException: Can't compare maps!
	at org.apache.avro.generic.GenericData.compare(GenericData.java:984)
	at org.apache.avro.specific.SpecificData.compare(SpecificData.java:333)
	at org.apache.avro.generic.GenericData.compare(GenericData.java:961)
	at org.apache.avro.specific.SpecificData.compare(SpecificData.java:333)
	at org.apache.avro.generic.GenericData.compare(GenericData.java:946)
	at org.apache.avro.specific.SpecificRecordBase.compareTo(SpecificRecordBase.java:81)
	at org.apache.avro.specific.SpecificRecordBase.compareTo(SpecificRecordBase.java:30)
	at org.apache.spark.sql.catalyst.encoders.ExpressionEncoderSuite.$anonfun$encodeDecodeTest$1(ExpressionEncoderSuite.scala:442)

```

Reply:
The issue here is how the `input` and the `convertedBack` objects are compared

if we replace the check, in line ExpressionEncoderSuite.scala:442,

`left.asInstanceOf[Comparable[Any]].compareTo(right) == 0`
By
`left.asInstanceOf[Comparable[Any]].equals(right) == 0`

the test for Avro encoder passes, but unfortunately other tests fail

Equality of objects is tricky.

The GenericData `compare` function 
```
  /** Comparison implementation.  When equals is true, only checks for equality,
   * not for order. */
  @SuppressWarnings(value="unchecked")
  protected int compare(Object o1, Object o2, Schema s, boolean equals) {
```
fails to compare Maps when the parameter `equals` is `false`


I propose the following

replace ExpressionEncoderSuite:434:444 lines

```
      val isCorrect = (input, convertedBack) match {
        case (b1: Array[Byte], b2: Array[Byte]) => Arrays.equals(b1, b2)
        case (b1: Array[Int], b2: Array[Int]) => Arrays.equals(b1, b2)
        case (b1: Array[Array[_]], b2: Array[Array[_]]) =>
          Arrays.deepEquals(b1.asInstanceOf[Array[AnyRef]], b2.asInstanceOf[Array[AnyRef]])
        case (b1: Array[_], b2: Array[_]) =>
          Arrays.equals(b1.asInstanceOf[Array[AnyRef]], b2.asInstanceOf[Array[AnyRef]])
        case (left: Comparable[_], right: Comparable[_]) =>
          left.asInstanceOf[Comparable[Any]].compareTo(right) == 0
        case _ => input == convertedBack
      }
```

by 

```
      val convertedBackRow = encoder.toRow(convertedBack)
      val isCorrect = row == convertedBackRow
```

With the proposed modification all the tests in ExpressionEncoderSuite passes

Reply:
I may be mistaken, we probably must compare `input` and `convertedBack` ; indeed what if the encoder creates a row missing fields, then the test of equality `row==convertedBackRow` does not guarantee that the encoder is correct.

I could not find a solution right now (I am still looking for an acceptable solution)